File: AtCoder/Browns/cpp/ARC114-A-Not_coprime.cpp
#include <bits/stdc++.h>
#define MAX_N 50
#define MAX_X 50
#define INF 1'000'000'000'000'000'000
using namespace std;
typedef long long ll;

ll N;
ll X[MAX_N];

vector<ll> primes() {
  vector<ll> res;
  vector<bool> is_prime(MAX_X + 1, true);
  for (ll i = 2; i <= MAX_X; ++i) {
    if (is_prime[i]) {
      res.push_back(i);
      for (ll j = i + i; j <= MAX_X; j += i) {
        is_prime[j] = false;
      }
    }
  }
  return res;
}

int main() {
  cin >> N;
  for (ll i = 0; i < N; ++i) {
    cin >> X[i];
  }
  vector<ll> ps = primes();
  ll res = INF;
  for (ll i = 1; i < (1 << ps.size()); ++i) {
    ll t = 1;
    for (ll j = 0; j < ps.size(); ++j) {
      if (i >> j & 1) t *= ps[j];
    }
    bool f = true;
    for (ll j = 0; j < N; ++j) {
      if (__gcd(t, X[j]) == 1) {
        f = false;
        break;
      }
    }
    if (f) res = min(res, t);
  }
  printf("%lld\n", res);
}

File: AtCoder/Browns/python/ARC114-A-Not_coprime.py
from math import gcd
MAX_X = 50
INF = float('inf')
N = int(input())
X = list(map(int, input().split()))


def primes():
    is_prime = [True] * (MAX_X + 1)
    res = []
    for i in range(2, MAX_X + 1):
        if is_prime[i]:
            res.append(i)
            for j in range(i + i, MAX_X + 1, i):
                is_prime[j] = False
    return res


ps = primes()
res = INF
for i in range(1, 1 << len(ps)):
    t = 1
    for j in range(len(ps)):
        if i >> j & 1:
            t *= ps[j]
    f = True
    for x in X:
        if gcd(x, t) == 1:
            f = False
            break
    if f:
        res = min(res, t)
print(res)
